### PR TITLE
Image gradient

### LINKED ISSUE #2319: Image gradient

### PR DESCRIPTION
**Pull request by mcquin**
Resolves #2319

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -341,6 +341,7 @@ def __compile_command(self):
             "pytest",
             "python-bioformats",
             "pyzmq",
+            "scikit-image",
```
Comment (0x00b1) [inline setup.py]: ?

File: cellprofiler/modules/imagegradient.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        y.pixel_data = y_data
+
+        images.add(y_name, y)
+
+        if self.show_window:
+            workspace.display_data.x_data = x_data
+
+            workspace.display_data.y_data = y_data
+
+    def display(self, workspace, figure):
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Missing two-dimensional display.

File: cellprofiler/modules/imagegradient.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        y = cellprofiler.image.Image(
+            dimensions=3,
+            parent_image=x
+        )
+
+        y.pixel_data = y_data
+
+        images.add(y_name, y)
+
+        if self.show_window:
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Missing `workspace.display_data.dimensions = …`.

File: cellprofiler/modules/imagegradient.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+        x_data = skimage.img_as_uint(x_data)
+
+        disk = skimage.morphology.disk(radius)
+
+        y_data = numpy.zeros_like(x_data)
+
+        for z, image in enumerate(x_data):
+            y_data[z] = skimage.filters.rank.gradient(image, disk)
+
+        y = cellprofiler.image.Image(
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Missing `dimensions`.

File: cellprofiler/modules/imagegradient.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+        x_data = x.pixel_data
+
+        x_data = skimage.img_as_uint(x_data)
+
+        disk = skimage.morphology.disk(radius)
+
+        y_data = numpy.zeros_like(x_data)
+
+        for z, image in enumerate(x_data):
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: I would expect a global gradient (i.e. use `gradient` from `skimage.filters`).

It has the added benefit of working on both two-and-three dimensional images.

Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Let’s revisit a local gradient in the future. I’ll make an issue.

File: cellprofiler/modules/imagegradient.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+
+        x = images.get_image(x_name)
+
+        x_data = x.pixel_data
+
+        x_data = skimage.img_as_uint(x_data)
+
+        disk = skimage.morphology.disk(radius)
+
+        y_data = numpy.zeros_like(x_data)
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Is this necessary?

File: cellprofiler/modules/imagegradient.py [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+
+        structure = self.structure.value
+
+        images = workspace.image_set
+
+        x = images.get_image(x_name)
+
+        x_data = x.pixel_data
+
+        x_data = skimage.img_as_uint(x_data)
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Is this necessary?

File: cellprofiler/modules/imagegradient.py [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+        self.x_name = cellprofiler.setting.ImageNameSubscriber(
+            "Input"
+        )
+
+        self.y_name = cellprofiler.setting.ImageNameProvider(
+            "Output",
+            "OutputImage"
+        )
+
+        self.structure = cellprofiler.setting.Choice(
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Use `cellprofiler.setting.StructuringElement`.

File: cellprofiler/modules/imagegradient.py [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+
+Image gradient
+
+"""
+
+import cellprofiler.image
+import cellprofiler.module
+import cellprofiler.setting
+import numpy
+import skimage.exposure
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Unused

File: cellprofiler/modules/imagegradient.py [hunk 10]
```diff
@@ -0,0 +1,102 @@
+"""
+
+Image gradient
```
Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Documentation?

Comment (0x00b1) [inline cellprofiler/modules/imagegradient.py]: Test?

File: setup.py (hunk 1 above)
Comment (mcquin) [inline setup.py]: We should declare it as a dependency, since it's used in modules. Correct?

File: cellprofiler/modules/imagegradient.py [hunk 11]
```diff
@@ -0,0 +1,98 @@
...
+
+        x_data = x.pixel_data
+
+        if x.dimensions is 3 or x.multichannel:
+            y_data = numpy.zeros_like(x_data)
+
+            for z, image in enumerate(x_data):
+                y_data[z] = skimage.filters.rank.gradient(image, self.__structuring_element())
+        else:
+            y_data = skimage.filters.rank.gradient(x_data, self.structuring_element.value)
```
Comment (mcquin) [inline cellprofiler/modules/imagegradient.py]: There is no non-local gradient filter in scikit-image: http://scikit-image.org/docs/dev/api/skimage.filters.html

**Comment by mcquin** [comment]
Ready for review. @0x00b1

**Pull request merged**